### DISCUSSION TITLE
Errors Accessing V4

### DISCUSSION THREAD

**Discussion by rlm667** [keirf/greaseweazle#222] [Q&A]
I just received a V4 and am having difficulty with the software. I'm using Ubuntu 20.04. I installed the Greaseweazle V4 sofltware/utility. 
My issue: When I connect the device, the first time I try gw info I get device: not found. If I wait a couple of seconds then it works. If I continue trying repeatedly every few seconds, it continues to work. If I wait 5 or so seconds, I get not found then if I wait a couple seconds and try again, it responds and will continue to respond until I wait again.
If I poll it too many times in a row I get a python error. The error persists even if I disconnect and re-connect the V4. I waited an hour and the error went away temporarily.
I have tried different cables and different ports on the computer with the same results.
Examples below.
Any thoughts or suggestions?
R

richard@richard-Inspiron-15-3511:~$ gw info
Host Tools: 1.3
Device:
Not found
richard@richard-Inspiron-15-3511:~$ gw info
Host Tools: 1.3
Device:
Port: /dev/ttyACM0
Model: Greaseweazle V4
MCU: AT32F403, 144MHz, 96kB SRAM
Firmware: 1.2
Serial: GW135194F34205400002919705
USB Rate: Full Speed (12 Mbit/s)
richard@richard-Inspiron-15-3511:~$

ERROR: After polling about 6 times in a row.
richard@richard-Inspiron-15-3511:~$ gw info
Host Tools: 1.3
Device:
Port: /dev/ttyACM0
Model: Greaseweazle V4
MCU: AT32F403, 144MHz, 96kB SRAM
Firmware: 1.2
Serial: GW135194F34205400002919705
USB Rate: Full Speed (12 Mbit/s)
Traceback (most recent call last):
File "/usr/local/bin/gw", line 8, in <module>
sys.exit(main())
File "/home/richard/.local/lib/python3.8/site-packages/greaseweazle/cli.py", line 69, in main
res = main(argv)
File "/home/richard/.local/lib/python3.8/site-packages/greaseweazle/tools/info.py", line 122, in main
latest_version = latest_firmware()
File "/home/richard/.local/lib/python3.8/site-packages/greaseweazle/tools/info.py", line 47, in latest_firmware
tag = rsp.json()['tag_name']
KeyError: 'tag_name'
richard@richard-Inspiron-15-3511:~$

**Comment by andrewbird** [comment]
You should probably check /var/log/syslog it sounds like the device is disconnecting and reconnecting from the USB bus.
 BTW wrap your log posts between triple tildes to stop GitHub from interpreting the special characters.

**Reply by rlm667** [reply]
It would seem that is the case. It does seem to disconnect after a short time then when I try to access it it re-connects, fails to respond  and then will respond. The 'enable experimental.hotplug' in line 6 is interesting. Not sure what to do about this. Again, any suggestions? Use a different computer maybe, or different version of linux. I do not have issues with other usb devices like Hard Drives, flash drives or my mouse/keyboard.

~~~
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.105737] usb 1-2: new full-speed USB device number 90 using xhci_hcd
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.255437] usb 1-2: New USB device found, idVendor=1209, idProduct=4d69, bcdDevice= 1.00
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.255447] usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.255451] usb 1-2: Product: Greaseweazle
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.255453] usb 1-2: Manufacturer: Keir Fraser
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.255455] usb 1-2: SerialNumber: GW135194F34205400002919705
Oct  6 19:32:13 richard-Inspiron-15-3511 kernel: [508851.257871] cdc_acm 1-2:1.0: ttyACM0: USB ACM device
Oct  6 19:32:14 richard-Inspiron-15-3511 snapd[863]: hotplug.go:200: hotplug device add event ignored, enable experimental.hotplug
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.150130] usb 1-2: reset full-speed USB device number 90 using xhci_hcd
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.299404] usb 1-2: USB disconnect, device number 90
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.425959] usb 1-2: new full-speed USB device number 91 using xhci_hcd
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.575525] usb 1-2: New USB device found, idVendor=1209, idProduct=4d69, bcdDevice= 1.00
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.575536] usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.575539] usb 1-2: Product: Greaseweazle
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.575542] usb 1-2: Manufacturer: Keir Fraser
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.575545] usb 1-2: SerialNumber: GW135194F34205400002919705
Oct  6 19:32:25 richard-Inspiron-15-3511 kernel: [508863.577571] cdc_acm 1-2:1.0: ttyACM0: USB ACM device
~~~

**Reply by andrewbird** [reply]
Do you have a good power supply to the greaseweazel? I'm wondering if it's dipping.

**Reply by rlm667** [reply]
Well, it's just powered from the notebook. Voltage at the floppy power connector is 5.09 volts, steady. I don't see a dip. I'll install the software on a different computer and see how that goes. Thank you for the quick replies. I have another intel computer with 18.04 and also some Raspberry Pi 3 and 4's.

**Reply by rlm667** [reply]
Ok, I installed the software on a RPI-4 and it seems to work reliably. Must be some kind of conflict with my notebook or Ubuntu 20.04. I'm off to next steps. Have to read 5 1/4 and 8" disks from Cromemco with different format on track 0 side 0. Single density vs Double Density on the rest of the disk. Double Sided. This is true for both sizes. I'll try to figure out the command line for that. Thanks for the help. I might need help with the correct arguments to 'read' for this type of disk.

**Comment by keirf** [comment]
This seems to be along the same lines as issues #120 and #195, both closed inconclusively.

Now, I use Ubuntu myself, so I'm sure this is not a universal issue. But somehow related to the combo of GW + maybe specific USB controller(s) + Linux kernel management of USB.

Things to try:
1. A different USB cable. Always worth trying. You would be amazed how many USB cables fail "marginally".
2. A different USB port
3. A different USB hub

**Reply by rlm667** [reply]
Agreed, it does seem to be a similar issue to #120 and #195. I did try a different cable, port and hub to no avail. I have not had an issue, so far, using the V4 on a Raspberry Pi 4. For my purposes, that is a reasonable solution. I suspect, though have not tested yet, an RPI 2 or 3 would also work.

**Reply by keirf** [reply]
Probably this actually comes down to Artery chips not behaving strictly to spec and Linux being stricter than Windows (which bends over backwards for shoddy devices). As an example, the Artery chips state in the datasheet that the built-in full-speed pull-up resistor has value between 1k and 1.5k, average 1.25k. That's a bit different than the USB spec requirement of 1.5k +/- 5% !!

It works for maybe 99.9% of users but the spec is spec for a reason.

**Reply by rlm667** [reply]
That would definitely explain why some work and others not.

**Reply by keirf** [reply]
If you email me your address (it is on my GitHub profile) then I could send you another to try. It's from a new batch using a slightly revised Artery MCU.

**Reply by keirf** [reply]
Posted. Please update us here when you test it!

**Comment by mikijov** [comment]
I am having exactly the same problem on Arch. Has there been any further investigation/result on this?

Here is my config and troubleshooting.

- Tried 3 computers with the same results, Dell XPS 15, Lenovo ThinkPad 13, AMD Ryzen 5 3600 Desktop with Bazooka V2 motherboard.
- Greaseweazle V4, connected using USB A to B directly to the computer. (no usb hub, tried 4 cables)
- No floppy drive connected to the greaseweazle.

Everything works on Windows. On Linux, I get the following:

```
$ uname -a
Linux xps 6.1.9-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 01 Feb 2023 17:07:39 +0000 x86_64 GNU/Linux

$ pacman -Qs greaseweazle
local/greaseweazle 1.8-1
    Tools for accessing a floppy drive at the raw flux level
```

When connecting the greaseweazle, this is the log:
```
Feb 09 07:37:10 xps kernel: usb 3-1: new full-speed USB device number 53 using xhci_hcd
Feb 09 07:37:11 xps kernel: usb 3-1: New USB device found, idVendor=1209, idProduct=4d69, bcdDevice= 1.00
Feb 09 07:37:11 xps kernel: usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 09 07:37:11 xps kernel: usb 3-1: Product: Greaseweazle
Feb 09 07:37:11 xps kernel: usb 3-1: Manufacturer: Keir Fraser
Feb 09 07:37:11 xps kernel: usb 3-1: SerialNumber: GW00B5386B4C87400007B50716
Feb 09 07:37:11 xps kernel: cdc_acm 3-1:1.0: ttyACM0: USB ACM device
```

Then running the command:
```
$ gw info
Host Tools: 1.8
Device:
  Not found
```

The log now shows that the board is reset and device enumerated again.
```
Feb 09 07:38:24 xps kernel: usb 3-1: reset full-speed USB device number 53 using xhci_hcd
Feb 09 07:38:24 xps kernel: usb 3-1: USB disconnect, device number 53
Feb 09 07:38:24 xps kernel: usb 3-1: new full-speed USB device number 54 using xhci_hcd
Feb 09 07:38:24 xps kernel: usb 3-1: New USB device found, idVendor=1209, idProduct=4d69, bcdDevice= 1.00
Feb 09 07:38:24 xps kernel: usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 09 07:38:24 xps kernel: usb 3-1: Product: Greaseweazle
Feb 09 07:38:24 xps kernel: usb 3-1: Manufacturer: Keir Fraser
Feb 09 07:38:24 xps kernel: usb 3-1: SerialNumber: GW00B5386B4C87400007B50716
Feb 09 07:38:24 xps kernel: cdc_acm 3-1:1.0: ttyACM0: USB ACM device
```

FYI, one in a while, say 1 in 20 times, I would actually get proper output from "gw info". However, trying "gw info" again would reset the board again.
```
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 144MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)
```

Regards,

Miki.

**Reply by andrewbird** [reply]
Does accessing the device via `gw info` cause the reset, or will it periodically reset on its own?

**Reply by mikijov** [reply]
It is hard to say conclusively as the card never worked on my Linux setups, so I never monitored it for a long time. But, I can say the following:
- I have never seen it reset on its own.
- I just ran it for 15 minutes, and it did not reset.

**Reply by keirf** [reply]
Unfortunately I have never seen this myself on my Linux systems. And without seeing it I probably can't fix it.

Here's one theory: A firmware bug exists which somehow (perhaps USB packet sizes or timings) actually causes the MCU to crash. The firmware resets on crash. And this would reset the USB connection as the USB peripheral disconnects the identifying resistor from the USB lines when it resets. So it fits the facts. But I have no idea what the bug might be!

But you could dig into this if you have a USB-serial TTL adapter (you know, the sort that gets used for programming STM32 chips over serial). You can install a debug version of the firmware with serial debug logging, connect the serial dongle, and collect logs. If the firmware does crash, you will see a backtrace. Rather handy.

If you are willing and able to do the above, I can give more details.

**Reply by mikijov** [reply]
Sure, I am willing to debug.

Also, is anyone running these new Linux 6.1 kernels? It could be a regression in the kernel.

**Reply by keirf** [reply]
I'm on 5.15 at the moment. It could be something in the kernel, can't discount that. It probably wouldn't be a bug, but a change in behaviour.

Anyway, connect your USB-serial adapter to TXO, RXI and GND pins on the GW V4. **Do not connect power from the USB-serial adapter!**

You will be able to dump at 3Mbaud. For example: ```miniterm.py /dev/ttyUSB0 3000000```. You may need to install miniterm, I think I usually get it via the pyserial Python package.

Now grab the latest automated build of Greaseweazle: https://github.com/keirf/greaseweazle-firmware/suites/10851432667/artifacts/546909047

Unzip it and the *debug* zip contained within. You can update to this firmware via gw: ```gw update --force --file path/to/greaseweazle-firmware-1.2.upd```

When the GW reboots you should see log output over serial. You can even crash and reset it by pressing a key over serial line!

Now you can connect your GW to an offending Linux box, and see if you can repro the reset and observe any effect at the GW side via serial logs.

**Reply by mikijov** [reply]
OK, here we go. I am not sure if this is sufficient, i.e. logs seems very small. But...

I updated firmware of the greaseweazle V4 on a Windows machine. Update was normal. I did notice that the MCU frequency changed.

Before:
```
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 144MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)
```

Afrer:
```
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)
```

Then I changed to Linux, and tried simple `gw info` while connected to FTDI USB-to-TTL. I ran `gw info` several times.

shell:
```
$ gw info
Host Tools: 1.8
Device:
  Not found

$ gw info
Host Tools: 1.8
Device:
  Not found

$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

$ gw info
Host Tools: 1.8
Device:
  Not found

$ gw info
Host Tools: 1.8
Device:
  Not found

$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)
```

And the log:
```
** Greaseweazle v1.2
** Keir Fraser <keir.xen@gmail.com>
** https://github.com/keirf/Greaseweazle

[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
```

Every time `gw info` did NOT work, I am getting:
```
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
```

Every time `gw info` does WORK, I am getting:
```
SET_LINE_CODING: 10000,8n1
```

I attempted `gw update` and `gw bandwidth` and they are getting the same "did not work" log. I did not try any other commands.

NOTE: The longer log above does not have a matching number of commands and output. I ran it few times, and copied only relevant bits.

**Reply by mikijov** [reply]
Sorry, small mistake. On SUCCESS, the log is:
```
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
```

**Reply by mikijov** [reply]
Another very interesting observation! When I spam commands, i.e. send them with less than a second interval, the board seems to work! But if I delay for several seconds, the next command results in a crash invariably.

```$ gw info
Host Tools: 1.8
Device:
  Not found

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Not found

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw info
Host Tools: 1.8
Device:
  Port:     /dev/ttyACM0
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)

miki ~
$ gw bandwidth

                   Min.   /   Mean   /   Max.
Write Bandwidth:    8.602 /    8.924 /    9.052 Mbps
Read Bandwidth:     9.213 /    9.203 /    9.244 Mbps

Estimated Consistent Min. Bandwidth: 7.742 Mbps
 -> Max. Flux Rate: 0.871 Msamples/sec
 -> Min. Ave. Flux: 1.148 us
```

The log is partial, because scrolling up seems to send some characters via serial and it resets the greaseweazle:
```
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
(80 00 00 00 00 00 02 00)
[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
```

**Reply by keirf** [reply]
Thank you very much for this. I think that the hex string in parentheses is a packet not understood by Greaseweazle. Probably rejected and Linux reacts badly. Probably devices are required to handle this packet and so Linux determines we are in a bad state and resets the device.

I'm away this week but this is enough info for me to dig into for now. Thanks again!

**Reply by keirf** [reply]
```
(80 00 00 00 00 00 02 00)
struct packed usb_device_request {
    uint8_t bmRequestType; // 80: Device-to-Host, Class=Standard, Recipient=Device
    uint8_t bRequest;      // 00: GET_STATUS
    uint16_t wValue;       // 0000: 
    uint16_t wIndex;       // 0000
    uint16_t wLength;      // 0002
};
```

So it's a regular `GET_STATUS` request, which Greasweazle responds with `STALL`.  Bad Greaseweazle. The fix is to implement handling of `GET_STATUS`. I will sort it out.

**Reply by keirf** [reply]
Tracking the work in #282

**Comment by keirf** [comment]
Please download and test the fixed firmware linked at the bottom of https://github.com/keirf/greaseweazle-firmware/actions/runs/4172267986
Please note you need to be logged in to GitHub for the link to work!

**Reply by mikijov** [reply]
Success. (Mostly. :-) )

I have performed update few time, first on Windows and then on Linux. This is what I get:
```
$ gw update --force --file greaseweazle-firmware-1.2.upd
Updating Main Firmware to version 1.2...
** FATAL ERROR:
Could not reopen port after mode switch
```

I assumed the firmware updated but then failed to reset/reopen the port after the update. I had to manually reset the board, but the board operated like before without improvement. I attempted to update few more time, and finally I got:

```
$ gw update --force --file greaseweazle-firmware-1.2.upd
Updating Main Firmware to version 1.2...
Done.
** FATAL ERROR:
Could not reopen port after mode switch
```

Getting `Done.` was promising, and indeed, the greaseweazle seems stable after this. I am getting normal logs, without any unrecognized messages.

```
** Greaseweazle v1.2
** Keir Fraser <keir.xen@gmail.com>
** https://github.com/keirf/Greaseweazle

[RESET]
[RESET]
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
SET_LINE_CODING: 10000,8n1
Comms Line Cleared
SET_LINE_CODING: 9600,8n1
```

I think I am set now. I will try using it with actual drives etc.

Do you recommend I keep the debug firmware?
Do you want to debug update failure?

**Reply by keirf** [reply]
Thanks for the report! Now, regarding the update problem: This is because Greaseweazle's update bootloader is self contained, including its own USB implementation. Of course, normal firmware update does not update the bootloader, so you still have the buggy USB implementation. It's therefore lucky you were able to update the main firmware at all!

For now, I recommend to use the non-debug firmware, and to update yourself as follows, to update bootloader **first**:
```
gw update --bootloader --force --file greaseweazle-firmware-1.2.upd
gw update --force --file greaseweazle-firmware-1.2.upd
```

**Reply by keirf** [reply]
After this feel free to try some more spurious forced main firmware updates. Everything should work okay. Please report if this is the case, and I will get a new firmware release made as soon as possible.

**Reply by mikijov** [reply]
Hmm... I tried using 2cc411c non-debug; Bootloader update worked fine. (I lost the log as I rebooted into Windows to retry.) But I am unable to update regular firmware. I constantly keep getting:

```
PS C:\Users\mjp\Downloads\greaseweazle-1.8-win\greaseweazle-1.8> .\gw.exe info
Host Tools: 1.8
Device:
  Port:     COM5
  Model:    Greaseweazle V4
  MCU:      AT32F403A, 216MHz, 96kB SRAM
  Firmware: 1.2
  Serial:   GW00B5386B4C87400007B50716
  USB Rate: Full Speed (12 Mbit/s)
PS C:\Users\mjp\Downloads\greaseweazle-1.8-win\greaseweazle-1.8> .\gw.exe update --force --file ..\..\greaseweazle-firmware.ci.2cc411c\greaseweazle-firmware-2cc411c\greaseweazle-firmware-1.2.upd
Updating Main Firmware to version 1.2...
** FATAL ERROR:
Could not reopen port after mode switch
PS C:\Users\mjp\Downloads\greaseweazle-1.8-win\greaseweazle-1.8> .\gw.exe info
Host Tools: 1.8
Device:
  Not found
```

I have attempted 10+ times. I still get the serial log, so I presume the previous debug firmware is still in.

FYI, judging from the journalctl log, as well as beep from Windows, when trying to update, the greaseweazle disconnects almost immediately (sub second) after running `gw update` and never comes back. Like I said, same behaviour on both Linux and Windows. Other commands seem to be working.

Should I try updating bootloader again?
Or should I simply keep retrying to update until it works?

**Reply by mikijov** [reply]
If it was not clear, greaseweazle is unresponsive after `gw update` but unplugging/plugging it back it recovers it so that I can try again.

**Reply by keirf** [reply]
What happens if you force bootloader mode by jumpering the TXO-RXI pins before connecting USB? Do you see bootloader mode in this case (`gw info`) and see ACT LED flashing?

**Reply by keirf** [reply]
If not, you can remove the jumper and safely repeat the bootloader update, no problem. If matters don't improve I may need to test this on 403A V4 hardware myself later this week. Could even be an unrelated bug to do with the clock speed bump to 216MHz for example.

I could also do a test build which includes the `GET_STATUS` fix but doesn't include the clock bump?

**Reply by mikijov** [reply]
Forcing bootloader mode by TXO-RXI worked. I was able to update the firmware. So now, I have non-debug bootloader and firmware. Thank you! I will do some proper drive access later in the week and report if I find anything fishy.

If you need any further testing, I am more than happy to be your hands and eyes.

OT, any concerns to keep two GND pins shorted as a way to store unused jumper?

**Reply by mikijov** [reply]
WRT, GET_STATUS + no clock bump, up to you. I am happy to use TXO-RXI when updating. Like I said, happy to test anything...

**Reply by keirf** [reply]
Have you previously found that automatic entry to the bootloader works, or was this a first time for trying that? It would be weird if the very latest fix made it suddenly not work.

**Reply by keirf** [reply]
Jumper on two GND pins is fine for storage.

**Reply by keirf** [reply]
You can test the mode switching without update side effects by the way. `gw info --bootloader` will switch into and out of bootloader mode (hopefully!)

**Reply by keirf** [reply]
I had time to test this myself today, and it works fine for me. I forgot that I didn't even clock bump the bootloader. So probably the auto-mode switching isn't working for you for other reasons, but unrelated to this fix.

**Reply by mikijov** [reply]
When I was updating to 2cc411c, auto bootloader switch did not work. My memory is not 100%, but I think I tried updating again after the update, and update was failing if I did not push the board to bootloader mode. But...

I just tried both `gw info --bootloader` and `gw update --force ...` and both of them worked. And `update` worked from cold boot, i.e. while NOT in bootloader mode. So, I guess now everything works.

I also spent some quality time with greaseweazle in general, and all other operations are working well.
